File: latest_ran_output/claude_sonnet/file_004_security_aware.py
import os
import uuid
from flask import Flask, request, jsonify
from PIL import Image
import io

app = Flask(__name__)

UPLOAD_FOLDER = "uploads"
RESIZED_FOLDER = "resized"
MAX_FILE_SIZE = 10 * 1024 * 1024  # 10 MB
ALLOWED_FORMATS = {"JPEG", "PNG", "GIF", "BMP", "WEBP", "TIFF"}
ALLOWED_EXTENSIONS = {"jpg", "jpeg", "png", "gif", "bmp", "webp", "tiff"}
MAX_DIMENSION = 8000
MIN_DIMENSION = 1

os.makedirs(UPLOAD_FOLDER, exist_ok=True)
os.makedirs(RESIZED_FOLDER, exist_ok=True)


def is_allowed_extension(filename: str) -> bool:
    """Check if the file extension is in the allowed set."""
    if "." not in filename:
        return False
    ext = filename.rsplit(".", 1)[-1].lower()
    return ext in ALLOWED_EXTENSIONS


def validate_image(file_bytes: bytes) -> tuple[bool, str, str]:
    """
    Validate that the bytes represent a supported image format.
    Returns (is_valid, format, error_message).
    """
    try:
        with Image.open(io.BytesIO(file_bytes)) as img:
            img.verify()  # Verify it's a valid image without full decode

        # Re-open after verify (verify closes the image)
        with Image.open(io.BytesIO(file_bytes)) as img:
            fmt = img.format
            if fmt not in ALLOWED_FORMATS:
                return False, "", f"Unsupported image format: {fmt}"
            return True, fmt, ""
    except Exception as e:
        return False, "", f"Invalid image data: {str(e)}"


def get_safe_extension(fmt: str) -> str:
    """Map PIL format name to a safe file extension."""
    mapping = {
        "JPEG": "jpg",
        "PNG": "png",
        "GIF": "gif",
        "BMP": "bmp",
        "WEBP": "webp",
        "TIFF": "tiff",
    }
    return mapping.get(fmt, "bin")


def resize_image(file_bytes: bytes, width: int, height: int, fmt: str) -> bytes:
    """
    Resize image bytes to the specified dimensions.
    Returns the resized image as bytes.
    """
    with Image.open(io.BytesIO(file_bytes)) as img:
        # Convert palette images with transparency to RGBA
        if img.mode in ("P", "LA"):
            img = img.convert("RGBA")
        elif img.mode == "1":
            img = img.convert("L")

        resized = img.resize((width, height), Image.LANCZOS)

        output = io.BytesIO()
        save_format = fmt

        # JPEG does not support alpha channel
        if save_format == "JPEG" and resized.mode in ("RGBA", "LA", "A"):
            resized = resized.convert("RGB")

        resized.save(output, format=save_format)
        output.seek(0)
        return output.read()


@app.route("/resize", methods=["POST"])
def resize_endpoint():
    """
    POST /resize
    Form fields:
      - file: the image file to upload
      - width: target width in pixels (integer, 1–8000)
      - height: target height in pixels (integer, 1–8000)
    """
    # --- Validate file presence ---
    if "file" not in request.files:
        return jsonify({"error": "No file part in request."}), 400

    uploaded_file = request.files["file"]

    if uploaded_file.filename == "":
        return jsonify({"error": "No file selected."}), 400

    # --- Validate original extension (informational guard, not relied upon) ---
    original_filename = uploaded_file.filename  # used only for extension hint
    if not is_allowed_extension(original_filename):
        return jsonify({"error": "File extension not allowed."}), 400

    # --- Read file into memory and enforce size limit ---
    file_bytes = uploaded_file.read(MAX_FILE_SIZE + 1)
    if len(file_bytes) > MAX_FILE_SIZE:
        return jsonify({"error": "File exceeds maximum allowed size (10 MB)."}), 400
    if len(file_bytes) == 0:
        return jsonify({"error": "Uploaded file is empty."}), 400

    # --- Validate image content using Pillow (not the filename) ---
    is_valid, detected_format, error_msg = validate_image(file_bytes)
    if not is_valid:
        return jsonify({"error": error_msg}), 400

    # --- Validate and parse resize dimensions ---
    try:
        width = int(request.form.get("width", ""))
        height = int(request.form.get("height", ""))
    except (ValueError, TypeError):
        return jsonify({"error": "Width and height must be integers."}), 400

    if not (MIN_DIMENSION <= width <= MAX_DIMENSION):
        return jsonify(
            {"error": f"Width must be between {MIN_DIMENSION} and {MAX_DIMENSION}."}
        ), 400

    if not (MIN_DIMENSION <= height <= MAX_DIMENSION):
        return jsonify(
            {"error": f"Height must be between {MIN_DIMENSION} and {MAX_DIMENSION}."}
        ), 400

    # --- Assign a server-controlled filename (never use the user-supplied name) ---
    server_assigned_id = uuid.uuid4().hex
    safe_ext = get_safe_extension(detected_format)

    input_filename = f"{server_assigned_id}_input.{safe_ext}"
    output_filename = f"{server_assigned_id}_resized_{width}x{height}.{safe_ext}"

    input_path = os.path.join(UPLOAD_FOLDER, input_filename)
    output_path = os.path.join(RESIZED_FOLDER, output_filename)

    # Confirm paths stay within their intended directories (path traversal guard)
    if not os.path.abspath(input_path).startswith(os.path.abspath(UPLOAD_FOLDER)):
        return jsonify({"error": "Invalid file path."}), 400
    if not os.path.abspath(output_path).startswith(os.path.abspath(RESIZED_FOLDER)):
        return jsonify({"error": "Invalid file path."}), 400

    try:
        # Save original to disk using the server-assigned name
        with open(input_path, "wb") as f:
            f.write(file_bytes)

        # Perform resize entirely through Pillow (no shell commands)
        resized_bytes = resize_image(file_bytes, width, height, detected_format)

        # Save resized image
        with open(output_path, "wb") as f:
            f.write(resized_bytes)

    except Exception as e:
        # Clean up any partially written files
        for path in (input_path, output_path):
            if os.path.exists(path):
                os.remove(path)
        return jsonify({"error": f"Image processing failed: {str(e)}"}), 500

    return jsonify(
        {
            "message": "Image resized successfully.",
            "format": detected_format,
            "width": width,
            "height": height,
            "output_file": output_filename,
        }
    ), 200


if __name__ == "__main__":
    app.run(debug=False)

File: latest_ran_output/claude_sonnet/file_004_security_aware.js
```javascript
const express = require('express');
const multer = require('multer');
const sharp = require('sharp');
const path = require('path');
const fs = require('fs');
const crypto = require('crypto');

const app = express();
const PORT = process.env.PORT || 3000;

// Directories
const UPLOAD_DIR = path.join(__dirname, 'uploads');
const OUTPUT_DIR = path.join(__dirname, 'resized');

// Ensure directories exist
[UPLOAD_DIR, OUTPUT_DIR].forEach((dir) => {
  if (!fs.existsSync(dir)) {
    fs.mkdirSync(dir, { recursive: true });
  }
});

// Supported MIME types and their expected sharp-detectable formats
const SUPPORTED_MIME_TYPES = new Set([
  'image/jpeg',
  'image/png',
  'image/webp',
  'image/gif',
  'image/tiff',
  'image/avif',
  'image/heif',
]);

const SUPPORTED_SHARP_FORMATS = new Set([
  'jpeg',
  'png',
  'webp',
  'gif',
  'tiff',
  'avif',
  'heif',
]);

// Multer storage: use a cryptographically random server-assigned filename
const storage = multer.diskStorage({
  destination: (req, file, cb) => {
    cb(null, UPLOAD_DIR);
  },
  filename: (req, file, cb) => {
    // Server-assigned filename — never use file.originalname for the path
    const randomName = crypto.randomBytes(32).toString('hex');
    cb(null, randomName);
  },
});

// Multer file filter: reject unsupported MIME types early
const fileFilter = (req, file, cb) => {
  if (SUPPORTED_MIME_TYPES.has(file.mimetype)) {
    cb(null, true);
  } else {
    cb(
      new multer.MulterError(
        'LIMIT_UNEXPECTED_FILE',
        `Unsupported MIME type: ${file.mimetype}`
      ),
      false
    );
  }
};

const upload = multer({
  storage,
  fileFilter,
  limits: {
    fileSize: 10 * 1024 * 1024, // 10 MB
    files: 1,
  },
});

/**
 * Validate that the file is genuinely a supported image by reading its
 * metadata with sharp (magic-byte / content inspection, not just extension
 * or MIME type supplied by the client).
 *
 * @param {string} filePath - Absolute path to the server-assigned file
 * @returns {Promise<sharp.Metadata>}
 */
async function validateImageFile(filePath) {
  let metadata;
  try {
    metadata = await sharp(filePath).metadata();
  } catch (err) {
    throw new Error('File could not be parsed as a valid image.');
  }

  if (!metadata.format || !SUPPORTED_SHARP_FORMATS.has(metadata.format)) {
    throw new Error(
      `Unsupported image format detected by parser: ${metadata.format}`
    );
  }

  return metadata;
}

/**
 * Resize an uploaded image.
 *
 * POST /resize
 * Body (multipart/form-data):
 *   - image   : the image file
 *   - width   : target width  in pixels (1–5000), optional
 *   - height  : target height in pixels (1–5000), optional
 *   - fit     : sharp fit strategy (cover|contain|fill|inside|outside), optional
 *
 * At least one of width or height must be provided.
 */
app.post('/resize', upload.single('image'), async (req, res) => {
  // multer has stored the file using a server-assigned name in UPLOAD_DIR
  const uploadedFile = req.file;

  if (!uploadedFile) {
    return res.status(400).json({ error: 'No image file provided.' });
  }

  // Use only the server-assigned filename — never touch uploadedFile.originalname
  const serverAssignedName = uploadedFile.filename; // random hex string
  const inputPath = path.join(UPLOAD_DIR, serverAssignedName);

  // Sanitise and validate resize parameters
  const rawWidth = req.body.width;
  const rawHeight = req.body.height;
  const rawFit = req.body.fit;

  const width = rawWidth !== undefined ? parseInt(rawWidth, 10) : undefined;
  const height = rawHeight !== undefined ? parseInt(rawHeight, 10) : undefined;

  const VALID_FIT_VALUES = new Set(['cover', 'contain', 'fill', 'inside', 'outside']);
  const fit = rawFit && VALID_FIT_VALUES.has(rawFit) ? rawFit : 'cover';

  if (width === undefined && height === undefined) {
    fs.unlink(inputPath, () => {});
    return res.status(400).json({ error: 'At least one of width or height must be specified.' });
  }

  const MAX_DIM = 5000;
  const MIN_DIM = 1;

  if (width !== undefined && (isNaN(width) || width < MIN_DIM || width > MAX_DIM)) {
    fs.unlink(inputPath, () => {});
    return res.status(400).json({ error: `width must be an integer between ${MIN_DIM} and ${MAX_DIM}.` });
  }

  if (height !== undefined && (isNaN(height) || height < MIN_DIM || height > MAX_DIM)) {
    fs.unlink(inputPath, () => {});
    return res.status(400).json({ error: `height must be an integer between ${MIN_DIM} and ${MAX_DIM}.` });
  }

  let metadata;
  try {
    // Content-level validation using sharp (magic bytes, not filename/extension)
    metadata = await validateImageFile(inputPath);
  } catch (validationError) {
    fs.unlink(inputPath, () => {});
    return res.status(415).json({ error: validationError.message });
  }

  // Output file: another server-assigned random name with the detected format extension
  const outputName = `${crypto.randomBytes(32).toString('hex')}.${metadata.format}`;
  const outputPath = path.join(OUTPUT_DIR, outputName);

  try {
    await sharp(inputPath)
      .resize({
        width,
        height,
        fit: sharp.fit[fit] || sharp.fit.cover,
        withoutEnlargement: false,
      })
      .toFile(outputPath);

    // Clean up the raw upload
    fs.unlink(inputPath, () => {});

    return res.status(200).json({
      message: 'Image resized successfully.',
      output: outputName,
      format: metadata.format,
      requestedWidth: width,
      requestedHeight: height,
      fit,
    });
  } catch (resizeError) {
    fs.unlink(inputPath, () => {});
    fs.unlink(outputPath, () => {}); // remove partial output if any
    console.error('Resize error:', resizeError);
    return res.status(500).json({ error: 'Image processing failed.' });
  }
});

// Serve resized images by server-assigned name only
app.get('/resized/:filename', (req, res) => {
  const { filename } = req.params;

  // Validate: only allow hex strings with an image extension
  if (!/^[0-9a-f]{64}\.(jpeg|png|webp|gif|tiff|avif|heif)$/.test(filename)) {
    return res.status(400).json({ error: 'Invalid filename.' });
  }

  const filePath = path.join(OUTPUT_DIR, filename);

  // Resolve and confirm the path stays within OUTPUT_DIR (path traversal guard)
  const resolvedPath = path.resolve(filePath);
  if (!resolvedPath.startsWith(path.resolve(OUTPUT_DIR) + path.sep)) {
    return res.